python/cuml/preprocessing/encoders.py

```diff
@@ -0,0 +1,291 @@
+        is present during transform (default is to raise). When this parameter
+        is set to 'ignore' and an unknown category is encountered during
+        transform, the resulting one-hot encoded columns for this feature
+        will be all zeros. In the inverse transform, an unknown category
+        will be denoted as None.
+
+    Attributes
+    ----------
+    drop_idx_ : array of shape (n_features,)
+        ``drop_idx_[i]`` is the index in ``categories_[i]`` of the category to
+        be dropped for each feature. None if all the transformed features will
+        be retained.
+    """
+    def __init__(self, categories='auto', drop=None, sparse=True,
+                 dtype=np.float, handle_unknown='error'):
+        self.categories = categories
+        self.sparse = sparse
+        self.dtype = dtype
+        self.handle_unknown = handle_unknown
+        self.drop = drop
+        self._fitted = False
+        self.drop_idx_ = None
+        self._encoders = None
+        if sparse and np.dtype(dtype) not in ['f', 'd', 'F', 'D']:
+            raise ValueError('Only float32, float64, complex64 and complex128 '
+                             'are supported when using sparse')
+
+    def _validate_keywords(self):
+        if self.handle_unknown not in ('error', 'ignore'):
+            msg = ("handle_unknown should be either 'error' or 'ignore', "
+                   "got {0}.".format(self.handle_unknown))
+            raise ValueError(msg)
+        # If we have both dropped columns and ignored unknown
+        # values, there will be ambiguous cells. This creates difficulties
+        # in interpreting the model.
+        if self.drop is not None and self.handle_unknown != 'error':
+            raise ValueError(
+                "`handle_unknown` must be 'error' when the drop parameter is "
+                "specified, as both would create categories that are all "
+                "zero.")
+
+    def _check_is_fitted(self):
+        if not self._fitted:
+            raise RuntimeError("Model must first be .fit()")
+
+    def _compute_drop_idx(self):
+        if self.drop is None:
+            return None
+        elif isinstance(self.drop, str) and self.drop == 'first':
+            return {feature: cp.array([0])
+                    for feature in self._encoders.keys()}
+        elif isinstance(self.drop, dict):
+            if len(self.drop.keys()) != len(self._encoders):
+                msg = ("`drop` should have as many columns as the number "
+                       "of features ({}), got {}")
+                raise ValueError(msg.format(len(self._encoders),
+                                            len(self.drop.keys())))
+            drop_idx = dict()
+            for feature in self.drop.keys():
+                cats = self._encoders[feature].classes_
+                self.drop[feature] = Series(self.drop[feature])
+                if len(self.drop[feature]) != 1:
```

Comment:
Little nitpicky, but error checks at the start of a new block make more sense to me

Reply:
I agree with you in general but in this case I use L123 (`self.drop[feature] = Series(self.drop[feature])`) to setup the variable before using it in error checks. L122 (`cats = self._encoders[feature].classes_`) is used the same way for the next error check below. The thing I can do though, is to move L122 closer to the second test something like that:
```python
L123
if test1:
  # do stuffs
L122
if test2:
```

Reply:
My point being here that you don't need to convert to `Series` to do `len`